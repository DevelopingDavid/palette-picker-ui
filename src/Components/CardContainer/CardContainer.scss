.cards-container {
  display: flex;
  flex-wrap: wrap;
}

.buttons-container {
  display: flex;
  flex-direction: column;
  align-items: center;
}

#invalid-name {
  color: red;
}
