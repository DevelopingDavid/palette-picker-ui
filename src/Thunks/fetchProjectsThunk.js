import { addProjects, isLoading, hasError } from '../Actions'

export const fetchProjectsThunk = () => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch('http://localhost:3001/api/v1/projects')
      if (!response.ok) {
        throw new Error(response.statusText)
      }
      const projects = await response.json()
      dispatch(isLoading(false))
      dispatch(addProjects(projects))
      return projects
    } catch (error) {
      dispatch(hasError(error.message))
    }
  }
}

export default fetchProjectsThunk