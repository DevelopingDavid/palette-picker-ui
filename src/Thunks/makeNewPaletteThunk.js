import { isLoading, hasError } from '../Actions'

export const makeNewPaletteThunk = (palette) => {
  return async (dispatch) => {
    const options = {
      method: 'POST',
      body: JSON.stringify(palette),
      headers: {
        "Content-Type": "application/json"
      }
    }
    try {
      dispatch(isLoading(true))
      const response = await fetch(`http://localhost:3001/api/v1/projects/${palette.project_id}/palettes`, options)
      if (!response.ok) {
        throw new Error(response.statusText)
      }
      dispatch(isLoading(false))
    } catch (error) {
      dispatch(hasError(error.message))
    }
  }
}

export default makeNewPaletteThunk;