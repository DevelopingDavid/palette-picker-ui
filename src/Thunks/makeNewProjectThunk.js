import { makeNewProject, isLoading, hasError } from '../Actions'

export const makeNewProjectThunk = (project) => {
  return async (dispatch) => {
    const options = {
      method: "POST",
      body: JSON.stringify(project),
      headers: {
        "Content-Type": "application/json"
      }
    }
    try {
      dispatch(isLoading(true))
      const response = await fetch(`${REACT_APP_BACKEND_URL}/api/v1/projects`, options)
      if (!response.ok) {
        throw new Error(response.statusText)
      }
      const projectId = await response.json()
      dispatch(isLoading(false))
      dispatch(makeNewProject(projectId))
      return projectId
    } catch (error) {
      dispatch(hasError(error.message))
    }
  }
}

export default makeNewProjectThunk