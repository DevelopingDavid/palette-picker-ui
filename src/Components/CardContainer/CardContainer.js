import React, { Component } from 'react'
import ColorBox from '../ColorBox/ColorBox'
import PropTypes from 'prop-types'
import Button from '@material-ui/core/Button'
import { connect } from 'react-redux'
import Modal from '@material-ui/core/Modal'
import { withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import makeNewProjectThunk from '../../Thunks/makeNewProjectThunk'
import makeNewPaletteThunk from '../../Thunks/makeNewPaletteThunk'
import fetchProjectsThunk from '../../Thunks/fetchProjectsThunk'
import shortid from 'shortid'
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';

const styles = theme => ({
  paper: {
    position: 'absolute',
    width: theme.spacing.unit * 50,
    backgroundColor: theme.palette.background.paper,
    boxShadow: theme.shadows[5],
    padding: theme.spacing.unit * 4,
    outline: 'none',
  },
  button: {
    margin: theme.spacing.unit,
  },
  input: {
    display: 'none',
  },
  formControl: {
    width: '60%'
  }
});

export class CardContainer extends Component {

  state = {
    open: false,
    projectName: '',
    selectedProjectId: '',
    invalidName: false
  };

  getModalStyle = () => {
    const top = 50
    const left = 50

    return {
      top: `${top}%`,
      left: `${left}%`,
      transform: `translate(-${top}%, -${left}%)`,
    };
  }

  handleOpen = () => {
    this.setState({ open: true });
  };

  handleClose = () => {
    this.setState({ open: false });
  };


  handleChange = (e) => {
    const { name, value } = e.target
    this.setState({
      [name]: value
    })
  }

  createNewProject = async () => {
    const { projects, makeNewProjectThunk } = this.props
    const newProject = {
      project_name: this.state.projectName,
    }
    const foundProject = projects.find(project => project.project_name === this.state.projectName)
    if (!foundProject) {
      this.setState({ invalidName: false })
      const id = await makeNewProjectThunk(newProject)
      this.postPalette(id)
    } else {
      this.setState({ invalidName: true })
    }
  }

  saveToExistingProject = () => {
    const id = this.state.selectedProjectId
    const { currentPalette, fetchProjectsThunk, makeNewPaletteThunk } = this.props
    const paletteWithId = {
      project_id: id,
      color_one: currentPalette[0].hex,
      color_two: currentPalette[1].hex,
      color_three: currentPalette[2].hex,
      color_four: currentPalette[3].hex,
      color_five: currentPalette[4].hex
    }
    makeNewPaletteThunk(paletteWithId)
    fetchProjectsThunk()
  }

  postPalette = async (id) => {
    const { fetchProjectsThunk, makeNewPaletteThunk } = this.props
    const projects = await fetchProjectsThunk()
    const foundId = await projects.find((project) => {
      return project.id === id.id
    })
    const { currentPalette } = this.props
    const paletteWithId = {
      project_id: foundId.id,
      color_one: currentPalette[0].hex,
      color_two: currentPalette[1].hex,
      color_three: currentPalette[2].hex,
      color_four: currentPalette[3].hex,
      color_five: currentPalette[4].hex
    }
    makeNewPaletteThunk(paletteWithId)
    fetchProjectsThunk()
  }

  handleSelectChange = event => {
    this.setState({ selectedProjectId: event.target.value });
  };

  render() {
    const { classes } = this.props;
    const { currentPalette } = this.props
    const makeCards = currentPalette.map(color => {
      return <ColorBox colorObject={color} key={shortid.generate()} />
    })
    
    const displayMenuItems = this.props.projects.map(project => {
      return <MenuItem key={shortid.generate()} value={project.id}>{project.project_name}</MenuItem>
    })

    return (
      <div className='container'>

        <div className='cards-container'>
          {makeCards}
        </div>
        <div className='buttons-container'>
          <Button variant="contained" color="primary" className={classes.button} onClick={this.props.checkLockedColors}>Generate New Colors</Button>
          <Button variant="contained" color="secondary" className={classes.button} onClick={this.handleOpen}>Save Palette</Button>
        </div>
        <Modal
          aria-labelledby="simple-modal-title"
          aria-describedby="simple-modal-description"
          open={this.state.open}
          onClose={this.handleClose}
        >
          <div style={this.getModalStyle()} className={classes.paper}>
            <Typography variant="h6" id="modal-title">
              Save New Project
            </Typography>
            <input name='projectName' value={this.state.projectName} onChange={this.handleChange} />
            <Button className='save-new-project-btn' onClick={this.createNewProject}>
              Save New Project
            </Button>
            <Typography variant="subtitle1" id="simple-modal-description">
              or
            </Typography>
            { this.state.invalidName ? <Typography variant="subtitle1" id="invalid-name">Sorry, that name is already in use. {<br />} Please choose a new name.</Typography> : undefined }
            <form className={classes.root} id='save-form' autoComplete="off">
              <FormControl className={classes.formControl}>
                <InputLabel htmlFor="age-simple">Select Existing Project</InputLabel>
                <Select
                  value={this.state.selectedProjectId}
                  onChange={this.handleSelectChange} >
                  { displayMenuItems }
                </Select>
                <Button className='save-to-existing-project-btn' onClick={this.saveToExistingProject}>
                  Save
                </Button>
              </FormControl>
            </form>
          </div>
        </Modal>
      </div>
    )
  }
}

CardContainer.propTypes = {
  makeNewProjectThunk: PropTypes.func.isRequired,
  currentPalette: PropTypes.array.isRequired,
  checkLockedColors: PropTypes.func.isRequired,
  projects: PropTypes.array
}

export const mapDispatchToProps = (dispatch) => ({
  makeNewProjectThunk: (project) => dispatch(makeNewProjectThunk(project)),
  makeNewPaletteThunk: (palette) => dispatch(makeNewPaletteThunk(palette)),
  fetchProjectsThunk: () => dispatch(fetchProjectsThunk())
})

export const mapStateToProps = state => ({
  currentPalette: state.currentPalette,
  projects: state.projects
})

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(CardContainer));