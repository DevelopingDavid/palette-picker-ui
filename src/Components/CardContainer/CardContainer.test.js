import React from 'react';
import { CardContainer, mapStateToProps, mapDispatchToProps } from './CardContainer';
import makeNewProjectThunk from '../../Thunks/makeNewProjectThunk';
import makeNewPaletteThunk from '../../Thunks/makeNewPaletteThunk';
import fetchProjectsThunk from '../../Thunks/fetchProjectsThunk';
import { createShallow } from '@material-ui/core/test-utils';
// import shortid from 'shortid';
// jest.mock('shortid');
// shortid.mockReturnValue("12345789");
jest.mock('../../Thunks/makeNewProjectThunk')
jest.mock('../../Thunks/makeNewPaletteThunk')
jest.mock('../../Thunks/fetchProjectsThunk')

let mockClasses = {
  root: '',
  paper: '',
  formControl: '',
  button: ''
};

const mockProjects = [
  {
    project_name: 'My Project',
    id: 50
  }
];

let mockPalette = [
  { hex: "#FB5985", locked: false },
  { hex: "#FFFFFF", locked: false },
  { hex: "#FFFFFF", locked: false },
  { hex: "#FFFFFF", locked: false },
  { hex: "#FFFFFF", locked: false }
];

describe('CardContainer', () => {
  let wrapper;
  let shallow;

  beforeEach(() => {
    shallow = createShallow();
    wrapper = shallow(
      <CardContainer
        classes={mockClasses}
        currentPalette={mockPalette}
        projects={mockProjects} />
    )
  });

  it.skip('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should return styles for modal', () => {
    expect(wrapper.instance().getModalStyle()).toEqual({
      top: `50%`,
      left: `50%`,
      transform: `translate(-50%, -50%)`,
    });
  });

  it('should handle open', () => {
    wrapper.instance().handleOpen();
    expect(wrapper.state('open')).toEqual(true);
  });

  it('should handle close', () => {
    wrapper.instance().handleClose();
    expect(wrapper.state('open')).toEqual(false);
  });

  it('should handle change', () => {
    const fakeEvent = { target: { name: 'projectName', value: 'oi' } };
    wrapper.instance().handleChange(fakeEvent);
    expect(wrapper.state('projectName')).toEqual('oi');
  });

  it('should change state for new entry', () => {
    const instance = wrapper.instance();
    // jest.spyOn(instance, 'postPalette');
    wrapper.setState({ projectName: 'New Project' });

    instance.createNewProject();

    expect(wrapper.state('invalidName')).toEqual(false);
    // expect(instance.postPalette).toHaveBeenCalled();
  });

  it('should change state for old entry', () => {
    const instance = wrapper.instance();
    wrapper.setState({ projectName: 'My Project' });

    instance.createNewProject();

    expect(wrapper.state('invalidName')).toEqual(true);
  });

  it('should change the state for id', () => {
    const fakeEvent = { target: { value: '1' } };
    wrapper.instance().handleSelectChange(fakeEvent);
    expect(wrapper.state('selectedProjectId')).toEqual('1');
  });

  describe('mapStateToProps', () => {
    it('should mapStateToProps', () => {
      const mockState = {}
      const upDatedState = {
        currentPalette: []
      }
      const firstState = mapStateToProps(mockState)
      expect(firstState).toEqual({})
      const newState = mapStateToProps(upDatedState)
      expect(newState).toEqual(upDatedState)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should dispatch makeNewProjectThunk', () => {
      const mockDispatch = jest.fn()
      const mockProject = {
        project_name: 'My Project'
      }
      const actionToDispatch = makeNewProjectThunk(mockProject)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.makeNewProjectThunk(mockProject)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should dispatch makeNewPaletteThunk', () => {
      const mockDispatch = jest.fn()
      const mockPalette = {
        project_name: 'a;lskdjf',
        color_one: '#fff',
        color_two: '#aaa',
        color_three: '#bbb',
        color_four: '#aaa',
        color_five: '#fff'
      }
      const actionToDispatch = makeNewPaletteThunk(mockPalette)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.makeNewPaletteThunk(mockPalette)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should dispatch fetchProjectsThunk', () => {
      const mockDispatch = jest.fn()
      const actionToDispatch = fetchProjectsThunk()
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.fetchProjectsThunk()
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})
