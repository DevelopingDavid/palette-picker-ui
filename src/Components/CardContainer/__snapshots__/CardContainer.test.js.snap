// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer should match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="cards-container"
  >
    <ConnectFunction
      colorObject={
        Object {
          "hex": "#FB5985",
          "locked": false,
        }
      }
      key="f2IfWUgvt"
    />
    <ConnectFunction
      colorObject={
        Object {
          "hex": "#FFFFFF",
          "locked": false,
        }
      }
      key="0Bu6JVet2z"
    />
    <ConnectFunction
      colorObject={
        Object {
          "hex": "#FFFFFF",
          "locked": false,
        }
      }
      key="lhbF9cPCGz"
    />
    <ConnectFunction
      colorObject={
        Object {
          "hex": "#FFFFFF",
          "locked": false,
        }
      }
      key="qqq8Asyyom"
    />
    <ConnectFunction
      colorObject={
        Object {
          "hex": "#FFFFFF",
          "locked": false,
        }
      }
      key="5ZIbkEWTxB"
    />
  </div>
  <div
    className="buttons-container"
  >
    <WithStyles(Button)
      className=""
      color="primary"
      variant="contained"
    >
      Generate New Colors
    </WithStyles(Button)>
    <WithStyles(Button)
      className=""
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      Save Palette
    </WithStyles(Button)>
  </div>
  <WithStyles(Modal)
    aria-describedby="simple-modal-description"
    aria-labelledby="simple-modal-title"
    onClose={[Function]}
    open={false}
  >
    <div
      className=""
      style={
        Object {
          "left": "50%",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        }
      }
    >
      <WithStyles(Typography)
        id="modal-title"
        variant="h6"
      >
        Save New Project
      </WithStyles(Typography)>
      <input
        name="projectName"
        onChange={[Function]}
        value=""
      />
      <WithStyles(Button)
        className="save-new-project-btn"
        onClick={[Function]}
      >
        Save New Project
      </WithStyles(Button)>
      <WithStyles(Typography)
        id="simple-modal-description"
        variant="subtitle1"
      >
        Please input the name of the Project you want to save.
      </WithStyles(Typography)>
      <form
        autoComplete="off"
        className=""
      >
        <WithStyles(FormControl)
          className=""
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="age-simple"
          >
            Select Existing Project
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(WithFormControlContext(Select))
            onChange={[Function]}
            value=""
          >
            <WithStyles(MenuItem)
              key="WReZ0oAx-E"
              value={50}
            >
              My Project
            </WithStyles(MenuItem)>
          </WithStyles(WithFormControlContext(Select))>
          <WithStyles(Button)
            className="save-to-existing-project-btn"
            onClick={[Function]}
          >
            Save
          </WithStyles(Button)>
        </WithStyles(FormControl)>
      </form>
    </div>
  </WithStyles(Modal)>
</div>
`;
